src/sources/socket/mod.rs

```diff
@@ -820,6 +823,45 @@ mod test {
         assert_eq!(source_result, Ok(()));
     }
 
+    #[tokio::test]
+    async fn tcp_connection_close_after_max_duration() {
+        let (tx, _) = SourceSender::new_test();
+        let addr = next_addr();
+
+        let mut source_config = TcpConfig::from_address(addr.into());
+        source_config.set_max_connection_duration_secs(Some(1));
+        let source_task = SocketConfig::from(source_config)
+            .build(SourceContext::new_test(tx, None))
+            .await
+            .unwrap();
+
+        // Spawn the source task and wait until we're sure it's listening:
+        let _ = tokio::spawn(source_task);
+        wait_for_tcp(addr).await;
+
+        let mut stream: TcpStream = TcpStream::connect(addr)
+            .await
+            .expect("stream should be able to connect");
+        let start = Instant::now();
+
+        let timeout = tokio::time::sleep(Duration::from_millis(1200));
+        let mut buffer = [0u8; 10];
+
+        tokio::select! {
+             _ = timeout => {
+                 panic!("timed out waiting for stream to close")
+             },
+             read_result = stream.read(&mut buffer) => {
+                 match read_result {
+                    // read resulting with 0 bytes -> the connection was closed
+                    Ok(0) => assert_relative_eq!(start.elapsed().as_secs_f64(), 1.0, epsilon = 0.1),
```

Comment:
🤔 
I ran the body of this test in a loop with a delay in between, and one time the threshold needed to be increased, so I set that to `0.2`. But twice now have hit a failure where the elapsed is a really small number like `0.00005`.

My main concern is introducing churn in CI failures on master merging this in as-is.

But I don't necessarily have a better proposal off the top of my head.

Reply:
Oh, good catch! I'll see if I can get to the bottom of that.

Reply:
I suppose it could be something that would never happen in CI since it wouldn't be looping over like this.
Naively I would expect it not to behave this way though 🤷

Reply:
Unfortunately, I'm not able to reproduce this. I ran the test on a loop for an hour and did not get a single failure. 

The diff is usually in the 0.0x range. A couple of times it went up to a ~0.18 diff, so I think 0.2 might still be a little bit on the low side. I think I'll bump that up to 0.3. 

But for the failures you've seen, I'm not sure how to proceed :(

Reply:
Thanks for taking a look at that. It could be something about how I'm looping. How are you invoking your loop?

Reply:
I changed the test to loop and just print out the result instead of asserting. See below:

```rust
  #[tokio::test]
  async fn tcp_connection_close_after_max_duration() {
      let (tx, _) = SourceSender::new_test();
      let addr = next_addr();

      let mut source_config = TcpConfig::from_address(addr.into());
      source_config.set_max_connection_duration_secs(Some(1));
      let source_task = SocketConfig::from(source_config)
          .build(SourceContext::new_test(tx, None))
          .await
          .unwrap();

      // Spawn the source task and wait until we're sure it's listening:
      let _ = tokio::spawn(source_task);
      println!("waiting for tcp to be ready");
      wait_for_tcp(addr).await;

      loop {
          let mut stream: TcpStream = TcpStream::connect(addr)
              .await
              .expect("stream should be able to connect");

          let start = Instant::now();
          let timeout = tokio::time::sleep(Duration::from_millis(1200));
          let mut buffer = [0u8; 10];

          tokio::select! {
               _ = timeout => {
                   println!("fail: timed out waiting for stream to close")
               },
               read_result = stream.read(&mut buffer) => {
                   match read_result {
                      // read resulting with 0 bytes -> the connection was closed
                      Ok(0) => {
                          let diff = start.elapsed().as_secs_f64() - 1.0;
                          if diff.abs() < 0.2 {
                              println!("okay: {}", diff)
                          } else {
                              println!("fail: unexpected connection duration: {}", diff);
                          }
                      }
                      Ok(_) => println!("fail: unexpectedly read data from stream"),
                      Err(e) => println!("fail:  {:}", e)
                   }
              }
          }
      }
  }
```

Reply:
Ah, that is the difference. I believe I was running the whole unit test case in a loop, albeit with some delay between runs.
I think we can disregard that line of inquiry.

`0.3` epsilon sounds good to me 👍

Reply:
Thanks @neuronull , I pushed the 0.3 change now :)